src/relay/op/nn/nn.cc

```diff
@@ -33,9 +34,22 @@ bool DenseRel(const Array<Type>& types,
 
   Array<tvm::Expr> oshape = data->shape;
   if (param->units.defined()) {
+    Array<tvm::Expr> dshape = data->shape;
+
+    // validate the weight shape is proper if defined
+    if (weight != nullptr) {
+      CHECK(reporter->AssertEQ(weight->shape[0], dshape[dshape.size() - 1]))
```

Comment:
weight != nullptr but units can still be not defined, we could dispatch by weight units is defined, the require weight to be not null when units is not defined